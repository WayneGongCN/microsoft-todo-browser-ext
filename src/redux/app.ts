import { createAsyncThunk, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { State } from ".";
import { authentication as msalAuthentication } from "../helpers/msal";
import request from "../helpers/request";
import { AuthenticationResult } from "@azure/msal-browser";
import { ICreateTaskParams, ITasklistResult } from "../types";


type SerializableAuthenticationResult = Omit<
  AuthenticationResult,
  "expiresOn"
> & { expiresOn: string };
export const authentication =
  createAsyncThunk<SerializableAuthenticationResult>(
    "app/authenticationResult",
    (_, { getState }) => {
      const state = getState() as State;
      const scopes = state.app.scopes;
      return msalAuthentication({ scopes }).then((res) =>
        JSON.parse(JSON.stringify(res))
      );
    }
  );


export const getTasklist = createAsyncThunk<ITasklistResult>(
  "app/getTasklist",
  (_, { getState }) => {
    const state = getState() as State;
    const headers = {
      Authorization: `Bearer ${state.app.authenticationResult.accessToken}`,
    };
    return request
      .get<ITasklistResult>("me/todo/lists", { headers })
      .then((res) => res.data);
  }
);

/**
 * Create task
 *
 * https://github.com/microsoftgraph/microsoft-graph-docs/blob/master/api-reference/v1.0/api/todotasklist-post-tasks.md
 */
export const createTask = createAsyncThunk<
  ITasklistResult,
  { tasklisId: string; task: ICreateTaskParams }
>("app/createTask", (params, { getState }) => {
  const state = getState() as State;
  const headers = {
    Authorization: `Bearer ${state.app.authenticationResult.accessToken}`,
  };
  return request
    .post(`me/todo/lists/${params.tasklisId}/tasks`, params.task, {
      headers,
    })
    .then((res) => res.data);
});

export const appSlice = createSlice({
  name: "app",

  initialState: {
    authenticationResult: null as null | SerializableAuthenticationResult,
    scopes: ["profile", "Tasks.ReadWrite"],
    tasklist: [] as ITasklistResult["value"],
  },

  reducers: {},

  extraReducers: (builder) => {
    builder
      .addCase(authentication.fulfilled, (state, action) => {
        state.authenticationResult = action.payload;
      })
      .addCase(getTasklist.fulfilled, (state, action) => {
        state.tasklist = action.payload.value;
      });
  },
});

export const asyncChunk = {
  authentication,
  getTasklist,
  createTask,
};
// @ts-ignore
appSlice.authentication = authentication;
// @ts-ignore
appSlice.getTasklist = getTasklist;
// @ts-ignore
appSlice.createTask = createTask;

export default appSlice;
