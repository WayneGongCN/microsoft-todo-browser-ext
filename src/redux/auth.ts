import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { State } from '.';
import { msalAcquireTokenSilent, msalAcquireTokenRedirect, msalGetAllAccounts } from '../helpers/msal';
import { AUTH_SCOPES } from '../constants';
import { bindAsyncActions } from '../helpers';
import AppError from '../helpers/error';
import { SerializAuthenticationResult } from '../../types';
import { ErrorCode } from '../constants/enums';

const SLICE_NAME = 'auth';
const initialState = {
  authenticationResult: null as null | SerializAuthenticationResult,
  loading: false,
  error: null as null | { code: ErrorCode; message: string },
};

/**
 * 静默获取 Token
 * - 登录态过期不会拉起登录弹窗
 */
const acquireTokenSilent = createAsyncThunk<SerializAuthenticationResult, void, { state: State; payload: SerializAuthenticationResult }>(
  `${SLICE_NAME}/acquireTokenSilent`,
  (_, { rejectWithValue }) => {
    const accounts = msalGetAllAccounts();
    if (accounts.length) {
      return msalAcquireTokenSilent({ scopes: AUTH_SCOPES, account: accounts[0] });
    } else {
      const err = new AppError({ code: ErrorCode.NOT_FOUND_ACCOUNT, message: 'Not found accounts' });
      rejectWithValue(err.serializ());
      return Promise.reject(err);
    }
  }
);

/**
 * 刷新 Token
 * - 登录态有效则直接返回
 * - 登录态过期则会拉起登录弹窗
 */
export const acquireTokenRedirect = createAsyncThunk<SerializAuthenticationResult, void, { state: State; payload: SerializAuthenticationResult }>(
  `${SLICE_NAME}/acquireTokenRedirect`,
  (_, { rejectWithValue }) =>
    msalAcquireTokenRedirect({ scopes: AUTH_SCOPES }).catch((e: AppError) => {
      rejectWithValue(e.serializ());
      return Promise.reject(e);
    })
);

/**
 * 获取 Token
 * - 如果有 Token 且没有过期，则直接返回 Token
 * - 如果有 Token 但已经过期，则调用 acquireTokenSilent 刷新 Token，刷新失败则调用 acquireTokenRedirect 重新登陆
 * - 如果没有 Token 则调用 acquireTokenRedirect 重新登陆
 */
export const acquireToken = createAsyncThunk<string, boolean | void, { state: State; payload: SerializAuthenticationResult }>(
  `${SLICE_NAME}/acquireToken`,
  (silent = false, { getState, dispatch, rejectWithValue }) => {
    const { auth } = getState();
    const { authenticationResult } = auth;
    const now = Date.now();
    const isExpires = now > (authenticationResult?.expiresOn || 0);

    if (isExpires && !silent) {
      return dispatch(acquireTokenSilent())
        .catch(() => dispatch(acquireTokenRedirect()))
        .then((res) => (res.payload as SerializAuthenticationResult).accessToken)
        .catch((e) => {
          rejectWithValue(e.serializ());
          return Promise.reject(e);
        });
    } else if (silent) {
      const err = new AppError({ code: ErrorCode.UNKNOW, message: 'Token expired but silent is true' });
      rejectWithValue(err.serializ());
      return Promise.reject(err);
    } else {
      return Promise.resolve(authenticationResult.accessToken);
    }
  }
);

export const authSlice = createSlice({
  name: SLICE_NAME,
  initialState,

  reducers: {},

  extraReducers: (builder) => {
    builder
      .addCase(acquireTokenRedirect.pending, (state) => {
        state.loading = true;
      })
      .addCase(acquireTokenRedirect.fulfilled, (state, { payload }) => {
        state.authenticationResult = payload;
        state.loading = false;
      })
      .addCase(acquireTokenRedirect.rejected, (state) => {
        state.loading = false;
      })

      // acquireTokenSilent
      .addCase(acquireTokenSilent.pending, (state) => {
        state.loading = true;
      })
      .addCase(acquireTokenSilent.fulfilled, (state, { payload }) => {
        state.authenticationResult = payload;
        state.loading = false;
      })
      .addCase(acquireTokenSilent.rejected, (state) => {
        state.loading = false;
      });
  },
});

export const asyncChunk = {
  acquireTokenRedirect,
  acquireTokenSilent,
};
bindAsyncActions(authSlice, asyncChunk);

export default authSlice;
