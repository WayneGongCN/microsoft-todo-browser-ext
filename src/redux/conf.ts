import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { persistReducer } from 'redux-persist';
import { State } from '.';
import { getConfRequest } from '../api/getConf';
import { getPersistConf } from '../helpers';



const SLICE_NAME = 'conf';
const APP_DEFAULT_CONF = {
  sentryDsn: 'https://365de3b2d6514a4caca2e06294ff74d0@o1093323.ingest.sentry.io/6112552',
  reportSampleRate: 0.1,
  gtmID: 'GTM-WVNS69V',
  gtmEvn: '',
};



export const fetchConfAction = createAsyncThunk<typeof APP_DEFAULT_CONF, void, { state: State }>(`${SLICE_NAME}/fetchConf`, () => {
  return getConfRequest()
});



const confSlice = createSlice({
  name: SLICE_NAME,


  initialState: {
    conf: APP_DEFAULT_CONF,
    loading: false,
  },


  reducers: {
    updateConf: (state, { payload }) => {
      state.conf = payload;
    },
  },


  extraReducers: (builder) => {
    builder
      .addCase(fetchConfAction.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchConfAction.fulfilled, (state, { payload }) => {
        state.conf = payload;
        state.loading = false;
      })
      .addCase(fetchConfAction.rejected, (state) => {
        state.loading = false;
      });
  },
});


const persistConfig = getPersistConf({ key: 'conf' })
export const persistConfReducer = persistReducer(persistConfig, confSlice.reducer)

export default confSlice;
