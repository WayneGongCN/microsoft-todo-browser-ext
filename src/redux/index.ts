import { AnyAction, configureStore } from '@reduxjs/toolkit';
import authSlice from './auth';
import { createLogger } from 'redux-logger';
import taskSlice from './task';
import tasklistSlice from './tasklist';
import popupSlice from './popup';
import { IS_DEV } from '../constants';
import optionsSlice from './options';
import { persistStore, persistReducer, FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import { onPresistReady } from '../helpers/persist';

const rootReducer = {
  auth: persistReducer({ key: 'auth', storage }, authSlice.reducer),
  task: taskSlice.reducer,
  tasklist: tasklistSlice.reducer,
  popup: popupSlice.reducer,
  options: persistReducer({ key: 'options', storage }, optionsSlice.reducer),
};

export const store = configureStore({
  devTools: IS_DEV,
  reducer: rootReducer,

  middleware: (getDefaultMiddleware) => {
    return getDefaultMiddleware({
      serializableCheck: {
        ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
      },
    }).concat([createLogger({ collapsed: true, duration: true })]);
  },
});

export const persistor = persistStore(store, undefined, onPresistReady);

export type State = ReturnType<typeof store.getState>;
export type Dispatch = typeof store.dispatch;
