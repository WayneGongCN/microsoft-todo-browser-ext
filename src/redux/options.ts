/* eslint-disable @typescript-eslint/ban-ts-comment */
import { createSlice } from '@reduxjs/toolkit';
import { createMigrate, persistReducer } from 'redux-persist';
import { ENABLE_DEBUG } from '../constants';
import { getPersistConf } from '../helpers';


export enum FillMode {
  TITLE = 'title',
  DESC = 'describe'
}

const APP_DEFAULT_OPTIONS = {
  enableKeepLastTasklist: true,
  defaultTasklistId: '',

  enableAutoFill: true,
  enableForceFill: true,
  fillMode: FillMode.TITLE,

  enableResetPopupForm: true,
  enableNotifacation: true,
  formSize: 'small' as 'medium' | 'small'
};


const optionsSlice = createSlice({
  name: 'options',

  initialState: {
    form: { ...APP_DEFAULT_OPTIONS },
    error: null,
  },

  reducers: {
    updateForm: (state, { payload }) => {
      state.form = { ...state.form, ...payload };
    },

    resetForm: (state) => {
      state.form = { ...APP_DEFAULT_OPTIONS };
    },
  },
});



const migrations = {
  133: (state: any) => {
    return {
      ...state,
      form: {
        enableKeepLastTasklist: true,
        defaultTasklistId: state.quickAddTaskTasklistId ?? '',

        enableAutoFill: state.enableQuickAdd ?? true,
        enableForceFill: true,
        fillMode: FillMode.TITLE,

        enableResetPopupForm: state.autoResetPopup ?? true,
        enableNotifacation: true,
        formSize: 'small',
      }
    }
  }
}


const persistConfig = getPersistConf({
  key: 'options',
  migrate: createMigrate(migrations, { debug: ENABLE_DEBUG }),
})
export const persistOptionsReducer = persistReducer(persistConfig, optionsSlice.reducer)

export default optionsSlice