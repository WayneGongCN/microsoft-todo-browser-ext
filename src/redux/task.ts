import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import { ICreateTaskParams, IPopupForm, ITaskResult } from '../../types';
import { TIME_ZONE } from '../constants';
import { ETaskContentTypes, ETaskImportance, NotifyType } from '../constants/enums';
import { LNAG_ADD_TASK, LNAG_OPEN_MS_TODO, LNAG_SUCCESS } from '../constants/lang';
import { bindAsyncActions, getActiveTab, openMicrosoftTodo } from '../helpers';
import Notify from '../helpers/notification';
import request from '../helpers/request';

export const mapToTaskParams = async (popupForm: IPopupForm, tab?: chrome.tabs.Tab): Promise<ICreateTaskParams> => {
  const { title, describe, bookmark, importance, dateTime } = popupForm;

  let bookmarkContent = '';
  if (bookmark) {
    const activeTba = tab || (await getActiveTab());
    bookmarkContent = `\n---\n${activeTba.title}\n${activeTba.url}`;
  }

  const content = `${describe}\n${bookmarkContent}`;
  const result: ICreateTaskParams = {
    title,
    body: {
      content,
      contentType: ETaskContentTypes.TEXT,
    },
    importance: importance ? ETaskImportance.HIGH : ETaskImportance.NORMAL,
  };

  if (dateTime) {
    result.reminderDateTime = {
      timeZone: TIME_ZONE,
      dateTime: dateTime,
    };
  }
  return result;
};

/**
 * 创建任务
 *
 * https://github.com/microsoftgraph/microsoft-graph-docs/blob/main/api-reference/v1.0/api/todotasklist-post-tasks.md
 */
export const createTask = createAsyncThunk<ITaskResult, IPopupForm>('task/createTask', async (params, { rejectWithValue }) => {
  const { tasklistId } = params;
  const data = await mapToTaskParams(params);
  return request
    .post<void, ITaskResult, ICreateTaskParams>(`me/todo/lists/${tasklistId}/tasks`, data)
    .then((res) => {
      new Notify({
        title: `${LNAG_ADD_TASK}${LNAG_SUCCESS}`,
        message: LNAG_OPEN_MS_TODO,
      })
        .onClick(() => openMicrosoftTodo(NotifyType.TASK, res.id))
        .show();
      return res;
    })
    .catch((e) => {
      rejectWithValue(e?.serializ());
      return Promise.reject(e);
    });
});

export const taskSlice = createSlice({
  name: 'task',

  initialState: {
    tasks: {} as Record<string, ITaskResult[]>,
    creating: false,
    error: null,
  },

  reducers: {},

  extraReducers: (builder) => {
    builder
      // createTask
      .addCase(createTask.pending, (state) => {
        state.creating = true;
      })
      .addCase(createTask.fulfilled, (state) => {
        state.creating = false;
      })
      .addCase(createTask.rejected, (state) => {
        state.creating = false;
      });
  },
});

export const asyncChunk = {
  createTask,
};
bindAsyncActions(taskSlice, asyncChunk);

export default taskSlice;
