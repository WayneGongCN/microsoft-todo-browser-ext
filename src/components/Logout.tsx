import React, { useCallback } from 'react';
import { LANG_LOADING_TEXT, LANG_LOGOUT_TEXT } from '../constants/lang';
import Button from '@material-ui/core/Button';
import CircularProgress from '@material-ui/core/CircularProgress';
import { backgroundContext } from '../helpers/background';
import { useDispatch, useSelector } from 'react-redux';
import { State } from '../redux';

const Logout: React.FC = () => {
  const dispatch = useDispatch();
  const loading = useSelector((state: State) => state.auth.loading);

  const handleLogout = useCallback(() => {
    const { authSlice } = backgroundContext;
    dispatch(authSlice.logout());
  }, []);

  return (
    <Button id="com-btn-logout" fullWidth disabled={loading} endIcon={loading ? <CircularProgress size={20} /> : null} onClick={handleLogout}>
      {loading ? LANG_LOADING_TEXT + '...' : LANG_LOGOUT_TEXT}
    </Button>
  );
};

export default Logout;
