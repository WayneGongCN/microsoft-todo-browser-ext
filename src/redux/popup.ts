import { createSlice } from '@reduxjs/toolkit';
import { DEFAULT_FORM_POPUP } from '../constants';
import { createTaskAction } from './task';
import { fetchTasklistAction } from './tasklist';



const popupSlice = createSlice({
  name: 'popup',

  initialState: {
    form: { ...DEFAULT_FORM_POPUP },
    creating: false,
    loadingTasklist: false,
    error: null,
  },

  reducers: {
    updateForm: (state, { payload }) => {
      state.form = { ...payload };
    },

    /**
     * 重置表单，tasklist 保持上次选中的内容
     */
    resetForm: (state) => {
      state.form = { ...DEFAULT_FORM_POPUP, tasklistId: state.form.tasklistId };
    },
  },

  extraReducers: (builder) => {
    builder
      // createTask
      .addCase(createTaskAction.pending, (state) => {
        state.creating = true;
      })
      .addCase(createTaskAction.fulfilled, (state) => {
        state.creating = false;
        state.form = { ...DEFAULT_FORM_POPUP, tasklistId: state.form.tasklistId };
      })
      .addCase(createTaskAction.rejected, (state) => {
        state.creating = false;
      })

      // fetchTasklist
      .addCase(fetchTasklistAction.pending, (state) => {
        state.loadingTasklist = true;
      })
      .addCase(fetchTasklistAction.fulfilled, (state, { payload }) => {
        state.loadingTasklist = false;
        if (!state.form.tasklistId) {
          const firstTasklist = payload.value[0];
          state.form.tasklistId = firstTasklist.id;
        }
      })
      .addCase(fetchTasklistAction.rejected, (state) => {
        state.loadingTasklist = false;
      });
  },
});

export default popupSlice;
