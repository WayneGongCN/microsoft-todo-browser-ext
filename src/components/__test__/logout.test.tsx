/* eslint-disable @typescript-eslint/ban-ts-comment */
import React from 'react';
import snapshotRender from 'react-test-renderer';
import { useSelector } from 'react-redux';
import { logoutAction } from '../../redux/auth';
import { cleanup, fireEvent, screen, render as domRender } from '@testing-library/react';
import Logout from '../Logout';


jest.mock('@azure/msal-browser');
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn(),
}));



describe('Snapshot test', () => {
  test('normal', () => {
    // @ts-ignore
    useSelector.mockReturnValue(false);
    const tree = snapshotRender.create(<Logout />).toJSON();
    expect(tree).toMatchSnapshot('normal');
  });

  test('loading', () => {
    // @ts-ignore
    useSelector.mockReturnValue(true);
    const tree = snapshotRender.create(<Logout />).toJSON();
    expect(tree).toMatchSnapshot('loading');
  });
});



jest.mock('../../redux/auth', () => ({ logout: jest.fn() }));
describe('Event test', () => {
  afterEach(cleanup);


  test('Click button', () => {
    // @ts-ignore
    useSelector.mockReturnValue(false);
    domRender(<Logout />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    // @ts-ignore
    expect(logoutAction.mock.calls.length).toEqual(1);
  });


  test('Click button on loading', () => {
    // @ts-ignore
    useSelector.mockReturnValue(true);
    domRender(<Logout />);

    const button = screen.getByRole('button');
    fireEvent.click(button);

    // @ts-ignore
    expect(logoutAction.mock.calls.length).toEqual(0);
  });
});
