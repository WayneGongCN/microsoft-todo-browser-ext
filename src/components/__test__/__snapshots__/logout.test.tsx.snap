// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test loading: loading 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text Mui-disabled MuiButton-fullWidth Mui-disabled"
  disabled={true}
  id="com-btn-logout"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={-1}
  type="button"
>
  <span
    className="MuiButton-label"
  >
    undefined...
    <span
      className="MuiButton-endIcon MuiButton-iconSizeMedium"
    >
      <div
        className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
        role="progressbar"
        style={
          Object {
            "height": 20,
            "width": 20,
          }
        }
      >
        <svg
          className="MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
            cx={44}
            cy={44}
            fill="none"
            r={20.2}
            strokeWidth={3.6}
            style={Object {}}
          />
        </svg>
      </div>
    </span>
  </span>
</button>
`;

exports[`Snapshot test normal: normal 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-fullWidth"
  disabled={false}
  id="com-btn-logout"
  onBlur={[Function]}
  onClick={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  />
</button>
`;
